Add a generic ItemPageLicenseFieldComponent

abelgomez opened the pull request:
## References

* Fixes #3170

## Description

This PR adds a new ItemPageLicenseFieldComponent, which allows showing generic rights information. If the component detects that the rights information points to a Creative Commons license, it delegates the rendering to the existing ItemPageCcLicenseFieldComponent.


## Instructions for Reviewers
Please add a more detailed description of the changes made by your PR. At a minimum, providing a bulleted list of changes in your PR is helpful to reviewers.


List of changes in this PR:
* First, ...
* Second, ...

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You do not need to complete this checklist prior creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [ ] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [ ] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [ ] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [ ] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [ ] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [ ] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [ ] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

abelgomez commented:
This PR should be ready for review.

Sorry for the long list of commits: when changing from the `dspace-8_x` branch to `main` I rebased rather than merging, and GitHub shows the commits inbetween.

Since this PR delegates the rendering of CC licenses to the ItemPageCcLicenseFieldComponent, this PR is affected by DSpace/DSpace#9882 and #3165.

I opted to keep ItemPageCcLicenseFieldComponent almost as is to avoid breaking any existing installation that relies on it.

tdonohue commented:
@abelgomez : Could you rebase your changes directly on `main`?  This long list of commits is problematic, as GitHub seems to think they are all part of this PR.  It's likely a problem on the branch that you created this PR from.   You may need to rebase that branch on latest `main` to solve the issue.

abelgomez commented:
As far as I remember, that was exactly the problem that caused the long list of commits: I rebased main onto my branch but GitHub does not detect that those commits were already on "main". I'll try to fix it, and if I fail, I'll close the PR and open a clean one.

abelgomez commented:
I started from scratch, and cherry-picked the interesting commits. Let me know if you want me to also combine several commits into one for a cleaner history.

tdonohue commented:
@abelgomez : If you wish this to be considered for 9.0, we'll need you to resolve the merge conflicts in this PR.  Our merger deadline for new feature PRs is March 28. So, any new feature PR not review, approved and merged by then will not be in 9.0.

tdonohue commented:
@abelgomez : our goal would be to *potentially* include all the PRs in DSpace 9.0.  However, every individual PR must be reviewed & approved separately.  So, I cannot make any guarantees which PRs would be included, as they cannot be included until they are reviewed & approved by at least one other developer.

If you want us to test the PRs in a particular *order*, we can do that.  But, it'd be best to specify which order you recommend they be tested in.  I started with testing https://github.com/DSpace/DSpace/pull/9882 alongside #3165 (as those seemed like the best place to start for me), but (as noted in [my review](https://github.com/DSpace/dspace-angular/pull/3165#pullrequestreview-2662671686)) I couldn't get them to work properly together.

abelgomez commented:
Sure! I understand that PRs must be tested 😊

I was asking about the feasibility and intentions (and the sentence "our goal would be to potentially include all the PRs in DSpace 9.0" has clarified it).

I'll check https://github.com/DSpace/dspace-angular/pull/3165, and I'll make suggestions on the order of the PRs once I get more insights.

alexandrevryghem reviewed (changes requested):
Thnx @abelgomez! I tested this functionality today and this works well! I added some minor feedback see comments below. I tested all the licence combinations and they still display correctly

While testing your new code I also saw that when you set the `cc.license.jurisdiction` to `nl` for example, the cc license that is being displayed is still the one from the US. Ideally the one with the euro sign should be displayed instead. The PNGs should probably also be replaced with [their matching SVG](https://creativecommons.org/mission/downloads/) at some point but this probably out of scope for this PR

abelgomez commented:
Hi @alexandrevryghem , I'll try to address all the issues ASAP, but I'm not sure I can do it before March 21.

abelgomez commented on src/app/item-page/simple/field-components/specific-field/cc-license/item-page-cc-license-field.component.ts:
```diff
@@ -60,14 +60,23 @@ export class ItemPageCcLicenseFieldComponent implements OnInit {
+   * @param uri
+   * @returns the CC code or null if uri is not a valid CC URI
+   */
+  public static parseCcCode(uri: string): string {
+    const regex = /.*creativecommons.org\/(licenses|publicdomain)\/([^/]+)/gm;
+    const matches = regex.exec(uri ?? '') ?? [];
+    return matches.length > 2 ? matches[2] : null;
+  }
```
Hi @alexandrevryghem , regarding this small piece of code, in fact I'm waiting for PR https://github.com/DSpace/dspace-angular/pull/3165 to be tested and merged to refactor it, since they touch the same parts of the code, and conflicts will probably arise.

Can you please jump to that PR and test it to see if it can be merged, and thus we can handle the conflicts and refactoring in this PR?

abelgomez commented on src/app/item-page/simple/field-components/specific-field/license/item-page-license-field.component.ts:
```diff
@@ -0,0 +1,104 @@
+  /**
+   * String to use as a separator if multiple rights entries are specified
+   */
+  @Input() separator = '•';
```
Done!

abelgomez commented on src/app/item-page/simple/field-components/specific-field/license/item-page-license-field.scss:
```diff
@@ -0,0 +1,4 @@
+.separator {
```
Done!

abelgomez commented on src/app/item-page/simple/field-components/specific-field/license/item-page-license-field.component.html:
```diff
@@ -0,0 +1,29 @@
+        </ng-container>
+        <ng-container *ngFor="let uri of uris; let last=last;">
+          <a [href]="uri" target="_blank" class="license-link"><span class="license-text">{{ uri }}</span></a>
+          <span class="separator" *ngIf="!last">{{ separator }}</span>
```
Done!

abelgomez commented:
Hi @alexandrevryghem ,

I have incorporated the changes you proposed.

This PR should be ready for review and merge. Keep in mind that, since I refactored the code to parse/validate CC URIs to a separate utility file as you suggested, this PR has conflicts with https://github.com/DSpace/dspace-angular/pull/3165.

Thus, if the plan is to include both, I'll need a small margin to reconcile both PRs when the first one is merged (whichever it will be).

abelgomez commented:
Ho @alexandrevryghem ,

as you wish. The point of this PR is to directly show the rights information in the simple page without requiring to modify the templates.

For me, seems weird that, by default, DSpaces show a badge and the license information if `dc.rights` and `dc.rights.uri` is set, but does not show anything if a digital library simply sets a "All rights reserved to the corresponding authors" (or whatever) in the `dc.rights` field.

Of course, institutions can customize what is shown or hidden by customizing templates, but this PR aims at showing relevant information for those who don't want to customize the templates.

Based on my experience, making big customizations to the Angular code makes upgrades difficult and time-consuming, since every major version typically includes several breaking changes, and institutions need to reconcile their templates with the "base" code. Thus, I think that some institutions would like to avoid as many customizations as possible.

Cheers!

abelgomez commented:
Hi @pilasou , 

Thanks for letting me know. This PR is not "abandoned", I plan to resolve the current conflicts and keep this PR in sync with the master branch and let the DSpace team decide whether it should be merged or not.

I simply haven't had the time to do it yet.

Cheers,

Abel

pilasou commented:
Thanks @abelgomez, no worries, i am not a developper so I cannot make code review, but once you resolve conflit, I will certainly test it!

EikLoe commented:
@abelgomez Thanks for your PR! I tested it locally and it worked as expected. 👍🏾
